src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs

```diff
@@ -788,5 +789,80 @@ public async Task BigWrite_SmallRead_Success(bool closeWithData)
                 }
             }
         }
+
+        [Fact]
+        public async Task BasicTest_WithReadsCompletedCheck()
+        {
+            QuicStream clientStream = null;
+            QuicStream serverStream = null;
+            await RunClientServer(
+                iterations: 100,
+                serverFunction: async connection =>
+                {
+                    QuicStream stream = await connection.AcceptStreamAsync();
+                    serverStream = stream;
+                    Assert.False(stream.ReadsCompleted);
+
+                    byte[] buffer = new byte[s_data.Length];
+                    int bytesRead = await ReadAll(stream, buffer);
+
+                    Assert.True(stream.ReadsCompleted);
+                    Assert.Equal(s_data.Length, bytesRead);
+                    Assert.Equal(s_data, buffer);
+
+                    await stream.WriteAsync(s_data, endStream: true);
+                    await stream.ShutdownCompleted();
+                },
+                clientFunction: async connection =>
+                {
+                    QuicStream stream = connection.OpenBidirectionalStream();
+                    clientStream = stream;
+                    Assert.False(stream.ReadsCompleted);
+
+                    await stream.WriteAsync(s_data, endStream: true);
+
+                    byte[] buffer = new byte[s_data.Length];
+                    int bytesRead = await ReadAll(stream, buffer);
+
+                    Assert.True(stream.ReadsCompleted);
+                    Assert.Equal(s_data.Length, bytesRead);
+                    Assert.Equal(s_data, buffer);
+
+                    await stream.ShutdownCompleted();
+                }
+            );
+
+            clientStream.Dispose();
+            serverStream.Dispose();
+        }
+
+        [Fact]
+        public async Task Read_ReadsCompleted_ReportedBeforeReturning0()
+        {
+            var sem = new SemaphoreSlim(0);
+
+            await RunBidirectionalClientServer(
+                async clientStream =>
+                {
+                    await sem.WaitAsync();
+                    await clientStream.WriteAsync(new byte[1], endStream: true);
+                },
+                async serverStream =>
+                {
+                    Assert.False(serverStream.ReadsCompleted);
+                    sem.Release();
+
```

Comment:
Should the test behave the same without the Semaphore? ReadsCompleted shouldn't return true until ReadAsync has returned all the data.

Reply:
Right, thanks for catching